aten/src/ATen/native/AutogradComposite.cpp

```diff
@@ -91,18 +93,72 @@ bool _has_same_storage_numel(const at::Tensor& base, const at::Tensor& other) {
   return base.storage().sym_nbytes() / base.itemsize() == other.storage().sym_nbytes() / other.itemsize();
 }
 
-Tensor _lazy_clone(Tensor const& self) {
+Tensor _lazy_clone(Tensor const& self, optional<c10::Device> device_opt) {
   c10::StorageImpl* self_storage = self.storage().unsafeGetStorageImpl();
-  c10::intrusive_ptr<c10::StorageImpl> storage =
-    c10::impl::cow::lazy_clone_storage(*self_storage);
+  c10::intrusive_ptr<c10::StorageImpl> storage = nullptr;
+
+  if (device_opt.has_value()) {
+    c10::Device src_device = self.device();
+    c10::Device dst_device = device_opt.value();
+    c10::DeviceType src_device_type = src_device.type();
+    c10::DeviceType dst_device_type = dst_device.type();
+    TORCH_CHECK(
+      (src_device_type == dst_device_type)
```

Comment:
Do you properly support different device index below? if not, there should be a check here

Reply:
I think different device indices would work, but we don't have any tests that exercise it right now (CI doesn't have any machines with two different MPS device indices). So it's probably best to raise an error to be safe